### PR TITLE
Explain lack of safeguarding access to provider users

### PR DESCRIPTION
**Pull request by mnacos**
### Context

Users who are unable to see safeguarding information due to it being forbidden by an org/org relationship need help to understand why they can't. This is a general need, but has become more pressing now we are onboarding greater numbers of providers, as their provider relationship permissions may not have been setup properly yet.

This PR introduces conditional text to explain the user's lack of access to safeguarding information for an application along the following lines (text provided by @adamsilver):

```
The candidate has disclosed sensitive material related to safeguarding.

{ if you don’t have permission but the relationship allows it }
   You cannot see safeguarding information.
   { if you can manage users }
       [Manage user permissions]
   { else you cannot manage users}
      [BLOCK FOR CONTACTS]
   { endif }
{ endif }

{ if you don’t have permission because the relationship forbids it }
  You have permission to see safeguarding information for the OAKS. However, the OAKS does not have permission to see safeguarding information for applications {ratified/run} by Blackfriars.
{ end }

{ if relationship not set }
    { if you can set them up }
        [Manage organisational permissions].
    { else if you can’t set them up }
        { if there are contacts }
            [BLOCK FOR CONTACTS]
        { else there aren’t any contacts }
             Email support at becomingateacher@digital.education.gov.uk
        { endif }
    { endif }
{ endif }
```

where BLOCK FOR CONTACTS is:

```
{% if just one contact %}
  Contact {{John Smith}} at {{john.smith@example.com}}.
{% else %}
  Contact one the following people:

  - {{John Smith}} at {{john.smith@example.com}}
  - {{Josh Brown}} at {{josh.brown@example.com}}
{% endif %}
```

UPDATE: After a lot of testing/discussion we have decided to avoid prompting the user to change anything if the org relationship has been set up. This includes dropping all 'Manage user permissions' links.

### Changes proposed in this pull request

Change `SafeguardingDeclarationComponent` to perform the relevant checks and display explanations according the current state of user-level permissions, org-level permissions and whether the provider relationship permissions have been set up.

Add a component preview for `SafeguardingDeclarationComponent` (`spec/components/previews/provider_interface/safeguarding_declaration_component_preview.rb`) to simplify development and testing for all these cases.

Add more providers and users to our `setup_local_dev_data` task and set up some of their permissions, so that the component previews have enough test data to work with.

### Guidance to review

The component preview depends heavily on the existence of appropriate provider relationships and user permissions to demonstrate all the different cases. The only way I could guarantee this on local and review environments was to add such cases to `setup_local_dev_data` and `GenerateTestApplications`.

You can inspect all different copy variations in action by visiting [the safeguarding component previews on the review app](https://apply-for-te-2876-expla-w3f3m3.herokuapp.com/rails/view_components/provider_interface/safeguarding_declaration_component), loading different cases and refreshing the browser (preview code will pick different cases at random with every page load). 

I've added an html debug comment to help with testing/reviewing, which we could remove before merging. On the other hand, there is no PII in this comment so we could leave as is. See `ProviderInterface::PermissionsDebugCommentComponent`.

image

This PR simply introduces a bit of text to a component, but the number of copy variations is extreme, as the copy depends on:

- user-level `view_safeguarding_information` permission of the `current_provider_user`,
- whether the course applied to is ratified by another organisation (that's when org-level restrictions come into play),
- whether `current_provider_user` is associated with the training or ratifying provider for the course the candidate has applied to,
- org-level `training_provider_can_view_safeguarding_information` permission of the training provider organisation or the ratifying provider equivalent (see previous point),
- whether the relevant org-level permission is `false` because this is the default or it has been actively set up this way,
- whether `current_provider_user` has `manage_organisations` or `manage_users` permission, if they are associated with the training provider for the course,
- and whether other users have `manage_organisations` or `manage_users` within the same organisation, if `current_provider_user` is a member of the training provider.

Have we covered all cases?

### Link to Trello card

[2876 - Show explanatory text to users who can't view safeguarding information due to org/org permissions](https://trello.com/c/gIujbsgQ)

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

**Comment by adamsilver** [comment]
1. Make sure text matches for org level permisions forbid viewing information but you belong to a ratifying provider
2. Show contact block for ratifying providers (and match content)
3. Make sure contact block uses classes for bullets

**Comment by mnacos** [comment]
> 1. Make sure text matches for org level permisions forbid viewing information but you belong to a ratifying provider
> 2. Show contact block for ratifying providers (and match content)
> 3. Make sure contact block uses classes for bullets

I've made these changes (and removed the word please) @adamsilver

**Comment by mnacos** [comment]
After a lot of testing with @adamsilver we have decided to avoid prompting the user to change anything if the org relationship has been set up. This includes dropping all 'Manage user permissions' links.

**Comment by duncanjbrown** [comment]
Is this correct? According to the block I have  view_safeguarding and manage_orgs at the training provider but I am being told the relationship forbids me to see this. Couldn't I change the relationship?

Screenshot 2020-11-04 at 09 30 35

**Comment by duncanjbrown** [comment]
Also here I'm the ratifying provider, but I'm being invited to set up permissions (or contact a colleague who can)

**Comment by mnacos** [comment]
> Is this correct? According to the block I have view_safeguarding and manage_orgs at the training provider but I am being told the relationship forbids me to see this. Couldn't I change the relationship?

The key factor in this case is the relationship has been set up, so even though the user has `manage_organisations` we don't prompt them to do so. Discussed this yesterday with @adamsilver

**Comment by mnacos** [comment]
> Also here I'm the ratifying provider, but I'm being invited to set up permissions (or contact a colleague who can)
> 
> Screenshot 2020-11-04 at 09 34 35

Ah, this is a bug I introduced last night, after simplifying the view. The expected behaviour here is prompting the current user to contact training provider admins or support. Fixing

**Comment by mnacos** [comment]
> Ah, this is a bug I introduced last night, after simplifying the view. The expected behaviour here is prompting the current user to contact training provider admins or support. Fixing

Fixed. Didn't notice because there is an edge case in the test data (provider user 2 vs. course 10) which is similar, but in this case the provider user belongs to both the ratifying _and_ the training provider.

**Review by adamsilver — APPROVED** [review]

File: lib/tasks/local_dev.rake [hunk 1]
```diff
@@ -35,14 +35,205 @@ task setup_local_dev_data: %i[environment copy_feature_flags_from_production syn
     make_decisions: true,
   )
 
+  puts 'Creating users for providers 1N1, 24P, 1JB, D39 with various permission combinations...'
+
+  # can_view__with_safeguarding_information
```
Comment (duncanjbrown) [inline lib/tasks/local_dev.rake:40]: This is all good and useful — maybe it could be tidier in a different file, though? `CreateExampleProviderUsersWithPermissions` or sth?

File: app/services/provider_authorisation_analysis.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
+class ProviderAuthorisationAnalysis
```
Comment (duncanjbrown) [inline app/services/provider_authorisation_analysis.rb:1]: This is nice and generic

**Review by duncanjbrown — COMMENTED** [review]
This is great, and I really like `ProviderAuthorisationAnalysis`. That file feels a bit like it's missing a spec. Maybe not at the unit level, but it would be good to have some sort of automated test for this?

File: app/services/provider_authorisation_analysis.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  attr_reader :permission, :auth, :application_choice, :course_option_id,
+              :training_provider, :ratifying_provider, :relationship
+
+  # This class is meant to be used after an auth call has already been performed
+  def initialize(permission:, auth:, application_choice:, course_option_id: nil)
```
Comment (benswannack) [inline app/services/provider_authorisation_analysis.rb:6]: Do we ever want to pass in a course option id separately?

Comment (mnacos) [reply app/services/provider_authorisation_analysis.rb:6]: Yes, when making or changing an offer an explicit `course_option_id` is required. So, if  `ProviderAuthorisationAnalysis` is used for `make_decisions` in this context, the course should be based on this `course_option_id`.

File: app/services/provider_authorisation_analysis.rb (hunk 2 above)
Comment (mnacos) [reply app/services/provider_authorisation_analysis.rb:1]: We'll be able to use this for `make_decisions` and `view_diversity_information` too, when https://github.com/DFE-Digital/apply-for-teacher-training/pull/3338 is merged.

File: app/services/provider_authorisation_analysis.rb (hunk 3 above)
Comment (mnacos) [reply app/services/provider_authorisation_analysis.rb:6]: See also: https://github.com/DFE-Digital/apply-for-teacher-training/pull/3273/commits/9c91f7fba93a1a3c2d03a80d2e02de2e4e80b4a1 (for `make_decisions`)

File: app/components/provider_interface/safeguarding_declaration_component.html.erb [hunk 4]
```diff
@@ -31,4 +34,70 @@
       <%= details %>
     </div>
   </details>
+<% elsif safeguarding_issues_declared? %>
+  <% if @analysis.provider_relationship_allows_access? %>
+    <p class="govuk-body">
+      The candidate has disclosed information to related to safeguarding but you do not have permission to view sensitive material.
```
Comment (duncanjbrown) [inline app/components/provider_interface/safeguarding_declaration_component.html.erb:40]:
```suggestion
      The candidate has disclosed information related to safeguarding but you do not have permission to view sensitive material.
```

Comment (mnacos) [reply app/components/provider_interface/safeguarding_declaration_component.html.erb]: Good spot

File: app/lib/create_example_provider_users_with_permissions.rb [hunk 5]
```diff
@@ -0,0 +1,192 @@
+class CreateExampleProviderUsersWithPermissions
```
Comment (duncanjbrown) [inline app/lib/create_example_provider_users_with_permissions.rb:1]: :+1:

**Review by duncanjbrown — APPROVED** [review]
🚢 !
